### PR TITLE
Update STRUMPACK solver interface

### PR DESCRIPTION
**Pull request by sebastiangrimberg**
Updates and upgrades for STRUMPACK sparse direct solver interface:

- HYPRE_BIGINT and HYPRE_MIXEDINT support.
- `Operator::ArrayMult(const Array<const Vector *> &X, Array<Vector *> &Y)` overrides for MUMPS solver classes (depends on https://github.com/mfem/mfem/pull/3075).
- Enable compression settings for recent versions of STRUMPACK.
- Mixed precision solver from STRUMPACK (requires fixes from STRUMPACK master branch above current release 6.3.1)
- Enable symbolic factorization reuse
- Fixed memory leak in `STRUMPACKRowLocMatrix` (`strumpack::CSRMatrixMPI` copies all inputs).

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3081](https://github.com/mfem/mfem/pull/3081) | @sebastiangrimberg | @v-dobrev | @psocratis + @sshiraiwa | 1/16/23 | 11/13/23 | 11/15/23 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by v-dobrev** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Comment by pghysels** [comment]
This looks very good.
I would perhaps add an option to enable/disable GPU off-loading in the factorization:
```
void STRUMPACKSolverBase<STRUMPACKSolverType>
::EnableGPU()
{
   solver_->options().enable_gpu();
}
void STRUMPACKSolverBase<STRUMPACKSolverType>
::DisableGPU()
{
   solver_->options().disable_gpu();
}
```
this doesn't require any other changes in the interface. GPU off-loading is available if STRUMPACK was compiled with CUDA, but input/output from MFEM to STRUMPACK is all still through host memory.
But any missing options can also be passed to the solver through the command line options (argc, argv).

Let me know if there are any other feature requests from STRUMPACK or this interface.

**Comment by sebastiangrimberg** [comment]
> I would perhaps add an option to enable/disable GPU off-loading in the factorization:

Done in 53bc105b. Also perhaps this is a question for @v-dobrev, but given that STRUMPACK v3 came out 5 years ago it might be fair to remove the `#if STRUMPACK_VERSION_MAJOR >= 3` checks in the code to clean things up? They don't do much harm I suppose so maybe leaving them is fine.

**Comment by v-dobrev** [comment]
> Also perhaps this is a question for @v-dobrev, but given that STRUMPACK v3 came out 5 years ago it might be fair to remove the `#if STRUMPACK_VERSION_MAJOR >= 3` checks in the code to clean things up? They don't do much harm I suppose so maybe leaving them is fine.

According to the INSTALL file: https://github.com/mfem/mfem/blob/e9eddf82056a0c129f580db035e2bc9ba4888eb7/INSTALL#L654-L662
the minimum required STRUMPACK version is 3.0.0, so there's no reason to check `#if STRUMPACK_VERSION_MAJOR >= 3`.

Also, as it turns out MFEM v4.5.2 has a small issue with STRUMPACK v6.3.1 and probably other versions before that, see #3593. We can address that here, I suppose, if it is still an issue after the changes in this PR.

**Comment by sebastiangrimberg** [comment]
Issue #3593 should be resolved by [417837e](https://github.com/mfem/mfem/pull/3081/commits/417837e238c20028711e1d55d7c75349a4551089). Thanks @v-dobrev!

File: linalg/strumpack.cpp [hunk 1]
```diff
@@ -345,7 +334,11 @@ FactorInternal() const
    strumpack::ReturnCode ret = solver_->factor();
    if (ret != strumpack::ReturnCode::SUCCESS)
    {
+#if STRUMPACK_VERSION_MAJOR >= 6 && STRUMPACK_VERSION_MINOR >= 3 && STRUMPACK_VERSION_PATCH > 1
```
Comment (v-dobrev) [inline linalg/strumpack.cpp:337]: This check is wrong: e.g. it will give false for v7.0.0 and v7.0.1.

I suspect there will be no new v6 version of STRUMPACK after v6.3.1, so the check can be simply
```c++
#if STRUMPACK_VERSION_MAJOR >= 7
```

Comment (sebastiangrimberg) [reply linalg/strumpack.cpp]: Whoops, good catch sorry about that. Working too quickly!

**Review by psocratis — COMMENTED** [review]
I am testing this on an M1 max and It seems that the Strumpack unit test hangs for # MPI procs >2. Can you check on your end @sebastiangrimberg

**Comment by pghysels** [comment]
I also see this hang. It looks like a problem in STRUMPACK. I'm looking into it.

**Comment by tzanio** [comment]
Thanks @pghysels

**Comment by sebastiangrimberg** [comment]
Yep, I am also able to reproduce this issue with STRUMPACK on my Intel Mac. Thanks a lot for your help @pghysels!

**Comment by pghysels** [comment]
I believe the hang in the unit test is fixed with this commit: https://github.com/pghysels/STRUMPACK/commit/627c9894c7ae7a0c96e6d671be81b05b01badbcf
I can make a new (7.1.2) bugfix release for this?

**Comment by sebastiangrimberg** [comment]
I can confirm that the latest `master` from STRUMPACK fixes the issue. Thanks for debugging this so quickly @pghysels!

**Comment by v-dobrev** [comment]
Thanks @pghysels!

Is it clear which STRUMPACK versions are affected by this issue? If it is clear, we can update the STRUMPACK version requirement to exclude those, here: https://github.com/mfem/mfem/blob/a53acb7ce814d05ca012201189ce097dbfc0ce0c/INSTALL#L661

**Comment by pghysels** [comment]
I believe all versions before 7.1.2 are affected.
But it is a corner case, it fails when the problem is too small to be distributed over the available MPI ranks.
You could also disable that unit test.

**Comment by v-dobrev** [comment]
> I believe all versions before 7.1.2 are affected. But it is a corner case, it fails when the problem is too small to be distributed over the available MPI ranks. You could also disable that unit test.

@sebastiangrimberg, can you try to increase the problem size in the unit test a bit so that older versions of STRUMPACK do not trigger this corner case? I hope the effect on runtime will be negligible.

**Comment by sebastiangrimberg** [comment]
> @sebastiangrimberg, can you try to increase the problem size in the unit test a bit so that older versions of STRUMPACK do not trigger this corner case? I hope the effect on runtime will be negligible.

Done: See [bdf529b](https://github.com/mfem/mfem/pull/3081/commits/bdf529b9fe49ee3ad5b6ce4f2372afdc562d70d2). Tested with v7.1.1 up to 8 processes.

**Comment by psocratis** [comment]
> > @sebastiangrimberg, can you try to increase the problem size in the unit test a bit so that older versions of STRUMPACK do not trigger this corner case? I hope the effect on runtime will be negligible.
> 
> Done: See [bdf529b](https://github.com/mfem/mfem/pull/3081/commits/bdf529b9fe49ee3ad5b6ce4f2372afdc562d70d2). Tested with v7.1.1 up to 8 processes. I think though we will always run into this corner case eventually based on how many cores someone running the test wants to use. As it is now we will fail for np > 8.

This still hangs for me only for the 1D case... (even though there is no empty processor)

**Comment by sebastiangrimberg** [comment]
> This still hangs for me only for the 1D case... (even though there is no empty processor). This is with STRUMPUCK 6.3.1 btw.

Hm, that is odd I'm not able to reproduce this. It hangs with <= 8 processes?

**Comment by sebastiangrimberg** [comment]
> This still hangs for me only for the 1D case... (even though there is no empty processor). This is with STRUMPUCK 6.3.1 btw.

Hi @psocratis, I was finally able to reproduce the hang with STRUMPACK v6.3.1 and have fixed it in 
1726be8. Can you test now? It should work with 8 processes now given the problem sizes (actual hang point for older versions of STRUMPACK is at `np=12`).

**Comment by pghysels** [comment]
When STRUMPACK uses SLATE for GPU support, it needs to call `MPI_Init_thread` with `MPI_THREAD_MULTIPLE`.
Same for when STRUMPACK uses PT-Scotch for ordering.
I don't see this in `ex11p` or `ex25p`, and it doesn't look like `mfem::Mpi` has a option for `MPI_Init_thread`.

**Comment by v-dobrev** [comment]
> When STRUMPACK uses SLATE for GPU support, it needs to call `MPI_Init_thread` with `MPI_THREAD_MULTIPLE`. Same for when STRUMPACK uses PT-Scotch for ordering. I don't see this in `ex11p` or `ex25p`, and it doesn't look like `mfem::Mpi` has a option for `MPI_Init_thread`.

Is there a STRUMPACK wrapper for MPI_Init that takes care of that? If not, is there a way to check if STRUMPACK (due to dependencies) needs `MPI_Init_thread` with `MPI_THREAD_MULTIPLE`?

**Comment by pghysels** [comment]
There is no wrapper for `MPI_Init(_thread)` in STRUMPACK.
In StrumpackConfig.h, there is 
```
#define STRUMPACK_USE_PTSCOTCH
#define STRUMPACK_USE_SLATE_SCALAPACK
```
or
```
/* #undef STRUMPACK_USE_PTSCOTCH */
/* #undef STRUMPACK_USE_SLATE_SCALAPACK */
```
and also (since STRUMPACK >= 7.1.0)
```
inline bool have_pt_scotch() { ...
inline bool have_slate() { ...
```

**Comment by v-dobrev** [comment]
Okay, thanks @pghysels, we should discuss if we want to add compile-time checks like this
```c++
#if defined(STRUMPACK_USE_PTSCOTCH) || defined(STRUMPACK_USE_SLATE_SCALAPACK)
// ...
#endif
```
in `mfem::Mpi::Init` or do something else. Any opinions?

**Comment by sebastiangrimberg** [comment]
> Okay, thanks @pghysels, we should discuss if we want to add compile-time checks like this
> 
> ```c++
> #if defined(STRUMPACK_USE_PTSCOTCH) || defined(STRUMPACK_USE_SLATE_SCALAPACK)
> // ...
> #endif
> ```
> 
> in `mfem::Mpi::Init` or do something else. Any opinions?

This would resolve the issue for STRUMPACK, but I'm worried about expecting MFEM to handle all cases where a dependency might be build in such a way that it requires `MPI_Init_thread` (with some value other than `MPI_THREAD_SINGLE`) instead of `MPI_Init`. MFEM has a lot of optional third-party library interfaces and having to update the `mfem::Mpi` wrapper any time one of these libraries makes a change expecting threaded MPI initialization might not be so scalable/maintainable.

I would think it is fine to keep `mfem::Mpi` as wrapping only `MPI_Init` and having the user required to call `MPI_Init_thread` in their application code based on how they are building dependencies. A more significant change might be to update `mfem::Mpi::Init` to take an argument that controls the threading level but this would involve changing most (or all) of the examples/miniapps to support this change.

In the case of Scotch/PT-Scotch dependencies, for example, building without `SCOTCH_PTHREAD_MPI` removes the need for `MPI_Init_thread` as I understand it. MUMPS may also use these dependencies in which case we would need to account for this in `mfem::Mpi::Init` to be consistent with STRUMPACK. Even worse, end users might use PT-Scotch as a replacement for ParMETIS in other dependencies which would mean the potential of having to handle all cases of checking for a linkage to threaded PT-Scotch in order for MFEM to initialize MPI correctly. My preference is to leave this to the user, but I understand there are pros/cons to both sides here.

**Comment by pghysels** [comment]
I think an MFEM user might not be aware of which dependency might need `MPI_THREAD_MULTIPLE`.
Ideally they run `spack install mfem` and get all the dependencies. Even for me working on STRUMPACK, it took me a while to figure out SLATE will behave erratically if not run with `MPI_THREAD_MULTIPLE`. 
I like Veselin's approach. But I know there can be performance implications, even for non-threaded code, from enabling `MPI_THREAD_MULTIPLE`. So alternatively `mfem::Mpi::Init` could just print a warning when `STRUMPACK_USE_SLATE_SCALAPACK` or `STRUMPACK_USE_PTSCOTCH` are defined saying the `MPI_THREAD_MULTIPLE` are required.

**Comment by sebastiangrimberg** [comment]
I like the warning idea. This supports the idea that we (MFEM) isn't promising to initialize MPI correctly for any other packages which are built requiring `MPI_THREAD_MULTIPLE`. See a357c4fe.

**Comment by v-dobrev** [comment]
I added a few small tweaks in 15934beb; from the commit message:
```
In the singleton class Mpi: added method Mpi::Init_thread that
performs MPI initialization using MPI_Init_thread; added method
Mpi::Init_auto that selects what MPI init mode to use based on
the configured external packages used by MFEM -- currently it
calls Mpi::Init_thread(MPI_THREAD_MULTIPLE) when using STRUMPACK
built with SLATE or PT-Scotch, otherwise it calls Mpi::Init.

In the examples that can use STRUMPACK, use Mpi::Init_auto instead
of Mpi::Init.
```
What do you think of these additions? I thought they will not be controversial, so I added them without checking with everyone. Let me know if there are objections.

I'm not sure I like the warning in `Mpi::Init` as it is now -- what if calling `Mpi::Init` is the correct way to initialize MPI in some setting but the method keeps printing the warning. We can add an option to `Mpi` to suppress such warnings. What do you think?

**Comment by sebastiangrimberg** [comment]
I think I’m still worried about trying to automatically determine which MPI initialization the user wants, though admittedly don't have a great solution. I definitely agree with providing the wrapper for `MPI_Init_thread` in the `Mpi` class, though we should expose the option for `MPI_THREAD_FUNNELED` which I think is used in most cases. Generally I’m concerned about committing that we can determine which MPI initialization the user wants in all cases, and feel it might lead to a bad user experience if we only commit to some cases or if we get it wrong.

For example, in the case of STRUMPACK + PT-Scotch, we only need `MPI_Init_thread` if PT-Scotch was so configured, which isn't always the case. So as it stands we are assuming PT-Scotch is always threaded. Likewise, we will need a separate case for MUMPS + PT-Scotch. There are bound to be other cases lurking which will come up as things mature and evolve. I am worried that this will start to create a long and deep chain of `if`/`else` for all dependencies which might be hard to maintain even if all dependencies can expose this information at compile time.

More and more I think this is a fair burden to put on the end user, and so the best approach may be to remove the “auto” variant from `Mpi`, and add command line options in the relevant examples which allow the user to control whether to use `Mpi::Init` or `Mpi::Init_thread` at runtime. I agree removing the warning from `Mpi::Init` would also be good and doable in this case. Let me know what you think about this plan, I’m happy to draft a commit for feedback if it sounds reasonable. I’m not sure, but it seems something might need to be added for the `punit_test` target as well.

**Comment by v-dobrev** [comment]
Just to clarify, `Mpi::Init_auto` is mostly meant for MFEM examples/miniapps where we know all the dependencies and can add the appropriate logic to handle as many cases as we want -- just to make sure examples/miniapps work out of the box. It can be used by users in their applications, however, if the logic there is not sufficient, users can make their own logic and use `Mpi::Init`, `Mpi::Init_thread`, or just standard MPI calls. Note that `Mpi::Init_thread` has the same parameters as `MPI_Init_thread` (excluding the `argc`+`argv` parameters which are typically unused), so using `MPI_THREAD_FUNNELED` is supported.

**Comment by tzanio** [comment]
Small suggestion: can we rename `Mpi::Init_auto()` to just `Mpi::Init()`?

**Comment by v-dobrev** [comment]
> Small suggestion: can we rename `Mpi::Init_auto()` to just `Mpi::Init()`?

`Mpi::Init` is already taken: the idea is that `Mpi::Init` is a wrapper for `MPI_Init` with automatic finalization and similarly, `Mpi::Init_thread` is a wrapper for `MPI_Init_thread` with automatic finalization. We can change `Mpi::Init_auto` to anything else but not `Mpi::Init`.

**Comment by sebastiangrimberg** [comment]
To @tzanio's point, we could have a single public method `Mpi::Init(int *argc = nullptr, char ***argv = nullptr, int requested = MPI_THREAD_SINGLE, int *provided = nullptr)`, since as far as I know `MPI_Init` is equivalent to `MPI_Init_thread` with `MPI_THREAD_SINGLE` in all MPI implementations.

Then we could move the STRUMPACK related `#ifdef`s into the relevant examples/tests? Or we could keep the `Mpi::Init_auto` (or `AutoInit` to keep the pascal case naming convention) wrapper in `Mpi`. I do still want to note that in the case of STRUMPACK + non-threaded PT-Scotch, the current set of proposed compile time checks will initialize MPI incorrectly and incur performance overhead.

**Comment by sebastiangrimberg** [comment]
PETSc has a global variable `PETSC_MPI_THREAD_REQUIRED` which has a default value and can be overridden by users before calling `PetscInitialize` (which calls `MPI_Init_thread`). Maybe we can add a public variable to the `Mpi` class, defaults set based on the STRUMPACK configuration variables, and just have this be the default parameter for `Mpi::Init`? That way no code changes are required in examples/miniapps, and users currently using the `Mpi` class can continue to do so and configure the threading level by modifying this variable, maybe `Mpi::thread_level_required`, at runtime.

**Comment by v-dobrev** [comment]
Hmm, it looks like we have a hang on Windows with MPI: https://github.com/mfem/mfem/actions/runs/5396076601/jobs/9817864609. I tried restarting the CI and got the same hang.

**Comment by v-dobrev** [comment]
@sebastiangrimberg, regading `mfem::Mpi`, can you summarize your proposed changes relative to the current state? I want to make sure I understand you latest proposal correctly.

**Comment by sebastiangrimberg** [comment]
Sure, here is the diff versus `HEAD~`:

```
diff --git a/general/communication.cpp b/general/communication.cpp
index 10fa6988c..62e46ffaf 100644
--- a/general/communication.cpp
+++ b/general/communication.cpp
@@ -26,6 +26,10 @@
 #include "sort_pairs.hpp"
 #include "globals.hpp"

+#ifdef MFEM_USE_STRUMPACK
+#include <StrumpackConfig.hpp> // STRUMPACK_USE_PTSCOTCH, etc
+#endif
+
 #include <iostream>
 #include <map>

@@ -34,6 +38,14 @@ using namespace std;
 namespace mfem
 {

+#if defined(MFEM_USE_STRUMPACK) && \
+    (defined(STRUMPACK_USE_PTSCOTCH) || defined(STRUMPACK_USE_SLATE_SCALAPACK))
+int Mpi::default_thread_required = MPI_THREAD_MULTIPLE;
+#else
+int Mpi::default_thread_required = MPI_THREAD_SINGLE;
+#endif
+
+
 GroupTopology::GroupTopology(const GroupTopology &gt)
    : MyComm(gt.MyComm),
      group_lproc(gt.group_lproc)
diff --git a/general/communication.hpp b/general/communication.hpp
index be8145689..5f2a0459e 100644
--- a/general/communication.hpp
+++ b/general/communication.hpp
@@ -32,10 +32,25 @@ namespace mfem
 class Mpi
 {
 public:
-   /// Singleton creation with Mpi::Init();
-   static void Init() { Init_(NULL, NULL); }
-   /// Singleton creation with Mpi::Init(argc,argv);
-   static void Init(int &argc, char **&argv) { Init_(&argc, &argv); }
+   /// Singleton creation with Mpi::Init(argc, argv).
+   static void Init(int &argc, char **&argv,
+                    int required = default_thread_required,
+                    int *provided = nullptr)
+   { Init(&argc, &argv, required, provided); }
+   /// Singleton creation with Mpi::Init().
+   static void Init(int *argc = nullptr, char ***argv = nullptr,
+                    int required = default_thread_required,
+                    int *provided = nullptr)
+   {
+      MFEM_VERIFY(!IsInitialized(), "MPI already initialized!");
+      int mpi_provided;
+      int mpi_err = MPI_Init_thread(argc, argv, required, &mpi_provided);
+      MFEM_VERIFY(!mpi_err, "error in MPI_Init()!");
+      if (provided) { *provided = mpi_provided; }
+      // The Mpi singleton object below needs to be created after MPI_Init() for
+      // some MPI implementations.
+      Singleton();
+   }
    /// Finalize MPI (if it has been initialized and not yet already finalized).
    static void Finalize()
    {
@@ -71,28 +86,19 @@ public:
    }
    /// Return true if the rank in MPI_COMM_WORLD is zero.
    static bool Root() { return WorldRank() == 0; }
+   /// Default level of thread support for MPI_Init_thread, override before calling Init().
+   static int default_thread_required;
 private:
-   /// Initialize MPI
-   static void Init_(int *argc, char ***argv)
+   /// Initialize the Mpi singleton.
+   static Mpi &Singleton()
    {
-      MFEM_VERIFY(!IsInitialized(), "MPI already initialized!")
-#if defined(MFEM_USE_STRUMPACK)
-#if defined(STRUMPACK_USE_PTSCOTCH) || defined(STRUMPACK_USE_SLATE_SCALAPACK)
-      if (Root())
-      {
-         MFEM_WARNING("STRUMPACK built with SLATE or PT-Scotch may require MPI_Init_thread with MPI_THREAD_MULTIPLE!");
-      }
-#endif
-#endif
-      MPI_Init(argc, argv);
-      // The "mpi" object below needs to be created after MPI_Init() for some
-      // MPI implementations
       static Mpi mpi;
+      return mpi;
    }
-   /// Finalize MPI
+   /// Finalize MPI.
    ~Mpi() { Finalize(); }
-   /// Prevent direct construction of objects of this class
-   Mpi() { }
+   /// Prevent direct construction of objects of this class.
+   Mpi() {}
 };

 /** @brief A simple convenience class based on the Mpi singleton class above.
```

Also, I am not sure if this was intentional, but it looks like https://github.com/mfem/mfem/commit/15934bebb97f74562c3b41d1464bd4ee1270547a introduced some unrelated (unintended?) changes to `miniapps/multidomain/multidomain.cpp`.

**Comment by sebastiangrimberg** [comment]
> > Also, I am not sure if this was intentional, but it looks like [15934be](https://github.com/mfem/mfem/commit/15934bebb97f74562c3b41d1464bd4ee1270547a) introduced some unrelated (unintended?) changes to `miniapps/multidomain/multidomain.cpp`.
> 
> It was intentional: from the commit message:
> 
> ```
> ...
> 
> Fix a small unrelated issue noticed during testing: in
> miniapps/multidomain/multidomain.cpp, do not open GLVis socket
> connections when visualization is disabled.
> ```

Whoops I'm sorry about that, I did not read far enough. Anyway I just pushed the changes in [77810e3](https://github.com/mfem/mfem/pull/3081/commits/77810e3741bd7e425a0e27644ed6d877927881ba).

**Comment by v-dobrev** [comment]
@sebastiangrimberg, the static variable `default_thread_required` probably needs `MFEM_EXPORT`, see https://github.com/mfem/mfem/blob/75d9ac43643feb9039133c25fe83cefcde6ced77/config/config.hpp#L37-L57

The workaround you found looks fine too but just using `MFEM_EXPORT` may simplify things a bit -- up to you if you want to modify things.

File: examples/ex11p.cpp [hunk 2]
```diff
@@ -262,12 +262,13 @@ int main(int argc, char *argv[])
 #ifdef MFEM_USE_STRUMPACK
       if (sp_solver)
       {
-         STRUMPACKSolver * strumpack = new STRUMPACKSolver(argc, argv, MPI_COMM_WORLD);
+         STRUMPACKSolver * strumpack = new STRUMPACKSolver(MPI_COMM_WORLD, argc, argv);
```
Comment (v-dobrev) [inline examples/ex11p.cpp:265]: This seems like a breaking change. Maybe we can have both versions of the constructor, so we don't break backward compatibility.

Comment (sebastiangrimberg) [reply examples/ex11p.cpp:265]: Fixed in [f5d0658](https://github.com/mfem/mfem/pull/3081/commits/f5d0658223984645535d01d5a68b18d653624fee).

**Comment by sebastiangrimberg** [comment]
> @sebastiangrimberg, the static variable `default_thread_required` probably needs `MFEM_EXPORT`, see

Thanks @v-dobrev, I did not know this about Windows builds. Fixed in [f5d0658](https://github.com/mfem/mfem/pull/3081/commits/f5d0658223984645535d01d5a68b18d653624fee).

**Review by psocratis — APPROVED** [review]

**Comment by tzanio** [comment]
ping: @sshiraiwa

**Comment by tzanio** [comment]
Thanks for the nice [design suggestion](https://github.com/mfem/mfem/pull/3081#issuecomment-1612133758) @sebastiangrimberg!

**Comment by sebastiangrimberg** [comment]
Hi @pghysels, do you mind taking a final look at this PR and giving it approval or requesting any changes you see need for? Thanks!

**Review by pghysels — APPROVED** [review]

**Comment by v-dobrev** [comment]
Merged in `next` for testing...

**Comment by v-dobrev** [comment]
Re-merged in `next` for testing...

**Comment by tepperly** [comment]
For what it's worth, I applied this PR locally and tested it with LiDO. This PR completely resolves the issues reported in #3978 .
